### ISSUE TITLE
Ubuntu 22.04 - libcrypto.so.1.1 not found

### ISSUE THREAD

**Issue by terry2010** [journeyapps/node-sqlcipher#101]
![图片](https://user-images.githubusercontent.com/1849037/185057799-20ae7cab-fe47-4b06-be54-30132f58ef7c.png)

```
find node_modules/@journeyapps/sqlcipher/lib/binding/ -type f | xargs ldd

node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node)
node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node)
node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by node_modules/@journeyapps/sqlcipher/lib/binding/napi-v6-linux-x64/node_sqlite3.node)
        linux-vdso.so.1 =>  (0x00007fff069e8000)
        libcrypto.so.1.1 => not found
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fb70aa43000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fb70a741000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fb70a52b000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb70a30f000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fb709f42000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb70b0f4000)
```

**Comment by RyQuantum** [comment]
Yes, encounter the same error on Ubuntu 22.04. Anyone knows how to fix it?

**Comment by terry2010** [comment]
> Yes, encounter the same error on Ubuntu 22.04. Anyone knows how to fix it?

`force user  to manually install old openssl (openssl 1.1)  could fix this......`

because latest ubuntu only have openssl3.x , but I hope there could have a better way  to resolve this

**Comment by RyQuantum** [comment]
> > Yes, encounter the same error on Ubuntu 22.04. Anyone knows how to fix it?
> 
> `force user to manually install old openssl (openssl 1.1) could fix this......`
> 
> because latest ubuntu only have openssl3.x , but I hope there could have a better way to resolve this

I find one, no need to downgrade the openssl. I bypass the issue by just installing the missing part from here: https://askubuntu.com/questions/1403778/upgrading-to-ubuntu-22-04-causes-libcrypto-errors-apt-dpkg-broken

After installing the libssl1.1, I can compile and run my project without any issue.

**Comment by eyaldar** [comment]
Hi, any one found better solution for this problem?
libssl1.1 has an open CVE: [CVE-2020-1971](https://github.com/advisories/GHSA-whf2-mq76-2fhv)

Is it possible to compile the linux version with the latest OpenSSL that exists on Ubuntu?

**Comment by dlliu114** [comment]
At present, some systems are installed with openssl 3. This library requires openssl 1.1  .  Can't let users install openssl 1.1 to solve this problem。Can linux build openssl 1.1 into the library like mac and windows？

**Comment by LemonSky2015** [comment]
Find a libcrypto.so.1.1 online and put it under lib64

**Comment by DigitalValhalla** [comment]
What I ended up doing *using debian w/ node 21.7.2, OpenSSL 3.0.15

assume the node project that is using @journeyapps/sqlcipher is here: /usr/local/projectthatusessqlcipher.

first run npm -i @journeyapps/sqlcipher in that directory:

_cd /usr/local/projectthatusessqlcipher
npm -i @journeyapps/sqlcipher_

next ensure the packages required to build from source are installed:

_sudo apt install git
sudo apt install python3
sudo apt install make
sudo apt install gcc
sudo apt install g++
sudo apt install libssl-dev_

create some one-off temp directory to git clone and build from source from

_mkdir /usr/local/node-sqlcipher-source
cd /usr/local/node-sqlcipher-source
git clone https://github.com/journeyapps/node-sqlcipher
cd ./node-sqlcipher
npm install --build-from-source_

there'll be a long litany of warnings and so forth as below

![image](https://github.com/user-attachments/assets/89a1695a-b0c2-4de9-a576-3725599fc4b2)

when all is complete, the npm install message will appear.

![image](https://github.com/user-attachments/assets/9e773444-3afd-495d-9d8e-e1e879c75590)

from there, run npm audit fix

next, delete the @journeyapps/sqlcipher from the node_modules of the project that was using sqlcipher unsuccessfully due to the libcrypto.so.1.1 not found error:

_rm -r -f /usr/local/projectthatusessqlcipher/node_modules/@journeyapps/sqlcipher_ 

replace it with the version you just built from source in the previous steps:

_cp -r /usr/local/node-sqlcipher-source/node-sqlcipher /usr/local/projectthatusessqlcipher/node_modules/@journeyapps/sqlcipher_
